src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/SimpleJoinOrderingQueryOptimizer.java

```diff
@@ -41,14 +48,26 @@ public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimi
     }
 
     public PhysicalPlan optimizePlan( final PhysicalPlan plan ) throws QueryOptimizationException {
+        final PhysicalPlan[] optSubPlans = getOptimizedSubPlans(plan);
+
         if ( hasMultiwayJoinAsRoot(plan) ){
-            final PhysicalPlan[] subplans = getOptimizedSubPlans(plan);
-            return joinPlanOptimizer.determineJoinPlan(subplans);
+            return joinPlanOptimizer.determineJoinPlan(optSubPlans);
         }
-        else {
-        	// TODO incorrect!! There may be multiway joins inside the subplans of the given plan.
+        else if ( hasMultiwayUnionAsRoot(plan) ){
+            return new PhysicalPlanWithNaryRootImpl((NaryPhysicalOp) plan.getRootOperator(), optSubPlans);
+        }
```

Comment:
Similar to my previous comment about the other helper methods (which now you have removed as suggested), instead of using this new helper method, `hasMultiwayUnionAsRoot`, we can use the number of subplans to recognize this case. Doing so will make the final else-branch (with the exception) obsolete. Therefore, you can simple remove this else-if branch here and move the code from within this else-if branch to the final else-branch, i.e.:

```java
   // ...
   else {
      return new PhysicalPlanWithNaryRootImpl((NaryPhysicalOp) plan.getRootOperator(), optSubPlans);
   }
```

Reply:
Solved.